pkg/core/transaction/signer.go

```diff
@@ -86,3 +86,34 @@ func SignersToStackItem(signers []Signer) stackitem.Item {
 	}
 	return stackitem.NewArray(res)
 }
+
+// Copy creates a deep copy of the Signer.
+func (c *Signer) Copy() *Signer {
+	if c == nil {
+		return nil
+	}
+	cp := *c
+	if c.AllowedContracts != nil {
+		cp.AllowedContracts = make([]util.Uint160, len(c.AllowedContracts))
+		copy(cp.AllowedContracts, c.AllowedContracts)
+	}
+
+	if c.AllowedGroups != nil {
+		cp.AllowedGroups = make([]*keys.PublicKey, len(c.AllowedGroups))
```

Comment:
Consider this code:
```
--- a/pkg/crypto/keys/publickey_test.go
+++ b/pkg/crypto/keys/publickey_test.go
@@ -40,14 +40,18 @@ func TestEncodeDecodePublicKey(t *testing.T) {
 }
 
 func TestPublicKeys_Copy(t *testing.T) {
-       pubs := make(PublicKeys, 5)
-       for i := range pubs {
+       pubz := make([]*PublicKey, 5)
+       for i := range pubz {
                priv, err := NewPrivateKey()
                require.NoError(t, err)
-               pubs[i] = priv.PublicKey()
+               pubz[i] = priv.PublicKey()
        }
+       pubs := PublicKeys(pubz)
 
        cp := pubs.Copy()
+       var pubx = ([]*PublicKey)(cp)
+       require.Equal(t, pubz, pubx)
+
        priv, err := NewPrivateKey()
        require.NoError(t, err)
        cp[0] = priv.PublicKey()
```
Just cast/Copy/cast again, no need for this.

Reply:
I don't see anything changed here.

Reply:
added a separate commit with changing the test. https://github.com/nspcc-dev/neo-go/pull/3407/files#diff-617aae89a8bc5f83ec727408c2b9b72f783683dc1d04acd4b9a67c6cbf2331ae

Reply:
Sorry, I don't care much about this test. It was just an example. What I care about a bit more here is dropping code duplication from `(*Signer).Copy()`. The idea with `PublicKeys.Copy()` reuse is a good one, but my point is that you don't need to change `AllowedGroups` type for it. Just cast/copy/cast and don't do if/make/for/copy. Less code, less problems.